extensions/amp-story/1.0/amp-story.js

```diff
@@ -1284,17 +1328,52 @@ export class AmpStory extends AMP.BaseElement {
       return Promise.resolve();
     }
 
-    if (
-      isExperimentOn(this.win, 'amp-story-paywall-exp') &&
-      targetPage.isPaywallProtected()
-    ) {
-      if (this.storeService_.get(StateProperty.SUBSCRIPTIONS_DIALOG_STATE)) {
-        // Subscription dialog is already triggered.
-        return Promise.resolve();
-      }
-      this.storeService_.dispatch(Action.TOGGLE_SUBSCRIPTIONS_DIALOG, true);
+    const subscriptionsState = this.storeService_.get(
+      StateProperty.SUBSCRIPTIONS_STATE
+    );
+    if (this.isPaywallStory_()) {
+      if (
+        pageIndex >= PAYWALL_PAGE_INDEX &&
+        subscriptionsState !== SubscriptionsState.GRANTED
+      ) {
+        if (subscriptionsState === SubscriptionsState.UNKNOWN) {
+          return this.blockOnPendingSubscriptionsState_(
+            targetPageId,
+            direction
+          );
+        }
+
+        // Attempt to navigate to the locked pages after the paywall page.
+        if (pageIndex > PAYWALL_PAGE_INDEX) {
+          this.pageAfterGranted_ = targetPageId;
 
-      // TODO(#37285): add SubscriptionService to actually trigger the subscription dialog.
+          // Show the paywall immediately if the active page is the paywall page.
+          if (this.isOnPaywallPage_()) {
+            this.showSubscriptionsDialog_();
+            return Promise.resolve();
+          }
+
+          // Redirect to the paywall page.
+          return this.switchTo_(
+            this.pages_[PAYWALL_PAGE_INDEX].element.id,
+            direction
+          );
+        }
+
+        if (this.pageAfterGranted_) {
+          this.showSubscriptionsDialog_();
+        } else {
+          this.paywallTimeout_ = setTimeout(() => {
```

Comment:
nit: can you add a comment that this else block is if we are on sub page we wait 2 sec